compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -323,52 +324,192 @@ generateS390Compare(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::Mne
+      generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, swapValues);
+
+      // code for handling +0/-0 comparisons when operands are equal
+      generateS390LabelInstruction(cg, TR::InstOpCode::label, node, equalRegion);
+      if (node->getOpCode().isMax())
+         {
+         // For Max calls, checking if first operand is +0, then we are done, otherwise fall through for swap
+         generateRXEInstruction(cg, tdcOpcode, node, operand1, generateS390MemoryReference(0x800, cg), 0);  // operand1 is +0 ?
+         generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK4, node, cFlowRegionEnd);   // it is +0
+         }
+      else if (node->getOpCode().isMin())
+         {
+         // For Min calls, checking if first operand is not +0, then we are done, otherwise fall through for swap
+         generateRXEInstruction(cg, tdcOpcode, node, operand1, generateS390MemoryReference(0x400, cg), 0);  // operand1 is -0 ?
+         generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_MASK4, node, cFlowRegionEnd);
+         }
       }
 
-   //Move resulting operand to lhsReg as fallthrough for alternate Condition Code
-   generateRREInstruction(cg, moveRROp, node, lhsReg, rhsReg);
+   // Move resulting operand to operand1 as fallthrough for alternate Condition Code
+   generateS390LabelInstruction(cg, TR::InstOpCode::label, node, swapValues);
+   TR::InstOpCode::Mnemonic moveRROp = TR::InstOpCode::NOP;
+   if (isFloatingPointOp)
+      {
+      moveRROp = node->getOpCode().isDouble() ? TR::InstOpCode::LDR : TR::InstOpCode::LER;
+      }
+   else
+      {
+      moveRROp = node->getOpCode().isLong() ? TR::InstOpCode::LGR : TR::InstOpCode::LR;
+      }
+   generateRREInstruction(cg, moveRROp, node, operand1, operand2);
 
    TR::RegisterDependencyConditions* deps = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, 2, cg);
-   deps->addPostConditionIfNotAlreadyInserted(lhsReg, TR::RealRegister::AssignAny);
-   deps->addPostConditionIfNotAlreadyInserted(rhsReg, TR::RealRegister::AssignAny);
+   deps->addPostConditionIfNotAlreadyInserted(operand1, TR::RealRegister::AssignAny);
+   deps->addPostConditionIfNotAlreadyInserted(operand2, TR::RealRegister::AssignAny);
 
    generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionEnd, deps);
    cFlowRegionEnd->setEndInternalControlFlow();
 
-   node->setRegister(lhsReg);
-   cg->decReferenceCount(lhsNode);
-   cg->decReferenceCount(rhsNode);
+   node->setRegister(operand1);
+   cg->decReferenceCount(node->getChild(0));
+   cg->decReferenceCount(node->getChild(1));
+   return operand1;
+   }
+
+
+TR::Register *
+OMR::Z::TreeEvaluator::fpMinMaxVectorHelper(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   TR_ASSERT_FATAL(node->getNumChildren() >= 1 && node->getNumChildren() <= 2 && node->getOpCode().isFloatingPoint(),
```

Comment:
Now that I've taken a second look at this assertion, should it be `node->getNumChildren() == 2 && node->getOpCode().isFloatingPoint()` instead?  The call to `node->getSecondChild()` below would cause problems if `node->getNumChildren()` returned one.

Reply:
done. I also changed the branching version of the helper to use an additional branch instruction in the floating point path. This allows it to skip over some of the instructions that deal with NaNs and improves performance by about 4% (in addition to the ~17% improvement that was already in this PR). 
please review @r30shah @hzongaro

Reply:
Hmm, Addiing an additional branch may interfere with the prediction unit. I assume you are seeing 3% in the bumblebench, is that consistent ?

Reply:
yup it's consistent. I ran the bumblebench benchmark with multiple different seeds for the RNG, multiple times per seed for each version. The one with the extra instruction consistently gets a score of 96M, and with no extra a score of 93M. Both have an uncertainty of 0.0% or 0.1% most of the time, and never above 1%

Reply:
> Hmm, Addiing an additional branch may interfere with the prediction unit. I assume you are seeing 3% in the bumblebench, is that consistent ? 
> 
> Also if I am reading correctly, we would have branch like following, 
> ```
> BRC  returnFirstArgCond, done
> BRC  returnSecondArgCond, swapValues
> BRC COND_8, equalOperands
> ...
> ```
> 
> 

Extra branch instruction has been removed as discussed